### PR TITLE
Ring Watch

### PR DESCRIPTION
**Pull request by aspacca**
Adding support for `Watch` method in `Ring` (this will make `Ring` fulfill `UniversalClient` interface)

### CODE REVIEW

File: ring.go [hunk 1]
```diff
@@ -275,14 +275,14 @@ func (c *ringShards) Heartbeat(frequency time.Duration) {
 func (c *ringShards) rebalance() {
 	hash := newConsistentHash(c.opt)
 	var shardsNum int
+	c.mu.Lock()
```
Comment (vmihailenco) [inline ring.go]: Is this necessary? `hash` is a local var and `shard.IsUp` uses atomics to it should be safe to be used in multiple goroutines.

File: ring.go [hunk 2]
```diff
@@ -653,6 +653,39 @@ func (c *Ring) Close() error {
...
+		}
+	}
+
+	if len(shards) == 0 {
+		return fmt.Errorf("redis: Watch requires at least one shard")
+	}
+
+	if len(shards) > 1 {
+		for _, shard := range shards[1:] {
+			if shard.Client.String() != shards[0].Client.String() {
```
Comment (vmihailenco) [inline ring.go]: Can we use `shard.Client != shards[0].Client` or `shard.Client.Options() != shards[0].Client.Options()`? Relying on `String` does not look right...

File: ring.go (hunk 1 above)
Comment (aspacca) [reply ring.go]: the problem is on `for name, shard := range c.shards`, you read from `c.shards` while `Ring.Close` can set it to `nil`, so the race condition (you can check travis build fails for the first commit)

Comment (aspacca) [reply ring.go]: maybe I can only `RLock` around the range and then keep the full lock before writing to `c.hash`

File: ring.go (hunk 2 above)
Comment (aspacca) [reply ring.go]: I will use the first, anyway `String` is unique on host/port, that should be what we care only for, that's why I used it in the first place. maybe I'm missing something

**Pull request merged**